### PR TITLE
IGNITE-22282 REST/CLI for disaster recovery: restart-partitions

### PR DESCRIPTION
**Pull request by Phillippko**
https://issues.apache.org/jira/browse/IGNITE-22282

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/cli/src/integrationTest/java/org/apache/ignite/internal/cli/commands/recovery/restart/ItRestartPartitionsTest.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        assertErrOutputIsEmpty();
+        assertOutputContains("Restarting partitions with cleanup.");
+    }
+
+    @Test
+    public void testRestartSpecifiedPartitions() {
+        execute(CLUSTER_URL_OPTION, NODE_URL,
+                RECOVERY_TABLE_NAME_OPTION, QUALIFIED_TABLE_NAME,
+                RECOVERY_ZONE_NAME_OPTION, ZONE,
+                RECOVERY_PARTITION_IDS_OPTION, "1,2"
```
Comment (rpuch) [inline modules/cli/src/integrationTest/java/org/apache/ignite/internal/cli/commands/recovery/restart/ItRestartPartitionsTest.java:78]: Do we have a test that makes sure that only the requested partitions are restarted?

File: modules/cli/src/integrationTest/java/org/apache/ignite/internal/cli/commands/recovery/restart/ItRestartPartitionsTest.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        String nodeNames = CLUSTER.runningNodes()
+                .limit(initialNodes() - 1)
+                .map(IgniteImpl::name)
+                .collect(joining(","));
+
+        execute(CLUSTER_URL_OPTION, NODE_URL,
+                RECOVERY_TABLE_NAME_OPTION, QUALIFIED_TABLE_NAME,
+                RECOVERY_ZONE_NAME_OPTION, ZONE,
+                RECOVERY_PARTITION_IDS_OPTION, "1,2",
+                RECOVERY_NODE_NAMES_OPTION, nodeNames
```
Comment (rpuch) [inline modules/cli/src/integrationTest/java/org/apache/ignite/internal/cli/commands/recovery/restart/ItRestartPartitionsTest.java:96]: Do we have a test that makes sure that only the requested nodes are affected?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java [hunk 3]
```diff
@@ -246,9 +246,17 @@ public CompletableFuture<Void> resetPartitions(String zoneName, String tableName
      * @param zoneName Name of the distribution zone. Case-sensitive, without quotes.
      * @param tableName Fully-qualified table name. Case-sensitive, without quotes. Example: "PUBLIC.Foo".
      * @param partitionIds IDs of partitions to restart. If empty, restart all zone's partitions.
+     * @param purge If partitions should be cleaned.
```
Comment (rpuch) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java:249]: Please elaborate what 'cleaning' means

File: modules/cli/src/main/java/org/apache/ignite/internal/cli/call/recovery/restart/RestartPartitionsCall.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        command.setNodeNames(input.nodeNames());
+        command.setTableName(input.tableName().trim());
+        command.setZoneName(input.zoneName().trim());
+        command.setPurge(input.purge());
+
+        try {
+            client.restartPartitions(command);
+
+            return DefaultCallOutput.success(
+                    "Restarting partitions with" + (input.purge() ? ""  : "out") + " cleanup."
```
Comment (rpuch) [inline modules/cli/src/main/java/org/apache/ignite/internal/cli/call/recovery/restart/RestartPartitionsCall.java:54]: Does the response arrive when we *start* restarting, or when the restart has *already* been done? Judging by the code of the manager, I suspect the latter, but the message says 'restarting' like it was just started.

File: modules/cli/src/integrationTest/java/org/apache/ignite/internal/cli/commands/recovery/restart/ItRestartPartitionsTest.java (hunk 2 above)
Comment (Phillippko) [reply modules/cli/src/integrationTest/java/org/apache/ignite/internal/cli/commands/recovery/restart/ItRestartPartitionsTest.java:96]: In slack thread Ivan Bessonov said that for CLI / REST test we should just check that CLI / REST works and we get "success". Restart logic should be tested in DisasterRecoveryManager and it's expensive to duplicate this checks in every REST / CLI test - sometimes it's not trivial. 

IMHO we need unit tests, to check that CLI calls REST with correct parameters...

File: modules/cli/src/integrationTest/java/org/apache/ignite/internal/cli/commands/recovery/restart/ItRestartPartitionsTest.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        assertOutputIsEmpty();
+        assertErrOutputContains("Some nodes are missing: [" + unknownNode + "]");
+    }
+
+    @Test
+    public void testRestartPartitionsIllegalPartitionNegative() {
+        execute(CLUSTER_URL_OPTION, NODE_URL,
+                RECOVERY_TABLE_NAME_OPTION, QUALIFIED_TABLE_NAME,
+                RECOVERY_ZONE_NAME_OPTION, ZONE,
+                RECOVERY_PARTITION_IDS_OPTION, "0,5,-1,-10"
```
[the remark this answers is not in the document]
Comment (Phillippko) [reply modules/cli/src/integrationTest/java/org/apache/ignite/internal/cli/commands/recovery/restart/ItRestartPartitionsTest.java:149]: Actually it reports minimal negative or maximum exceeding the limit number. But it's a detail of implementation, so removed -1 from test case.

File: modules/cli/src/main/java/org/apache/ignite/internal/cli/call/recovery/restart/RestartPartitionsCall.java (hunk 4 above)
Comment (Phillippko) [reply modules/cli/src/main/java/org/apache/ignite/internal/cli/call/recovery/restart/RestartPartitionsCall.java:54]: Manager returns future, CLI command completes when the future is returned, not when it's finished

File: modules/cli/src/integrationTest/java/org/apache/ignite/internal/cli/commands/recovery/restart/ItRestartPartitionsTest.java [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+        assertErrOutputIsEmpty();
+        assertOutputContains("Restarting partitions without cleanup.");
+    }
+
+    @Test
+    public void testRestartAllPartitionsPurge() {
+        execute(CLUSTER_URL_OPTION, NODE_URL,
+                RECOVERY_TABLE_NAME_OPTION, QUALIFIED_TABLE_NAME,
+                RECOVERY_ZONE_NAME_OPTION, ZONE,
```
[the remark this answers is not in the document]
Comment (Phillippko) [reply modules/cli/src/integrationTest/java/org/apache/ignite/internal/cli/commands/recovery/restart/ItRestartPartitionsTest.java:65]: CLI framework verifies that parameters fit into command contract (checks exclusive, mandatory parameters, etc)

[the remark this answers is not in the document]
Comment (rpuch) [reply modules/cli/src/integrationTest/java/org/apache/ignite/internal/cli/commands/recovery/restart/ItRestartPartitionsTest.java:65]: But do we have a test that verifies this? :)

**Review by rpuch — COMMENTED** [review]
.

File: modules/cli/src/main/java/org/apache/ignite/internal/cli/call/recovery/restart/RestartPartitionsCall.java (hunk 4 above)
Comment (rpuch) [reply modules/cli/src/main/java/org/apache/ignite/internal/cli/call/recovery/restart/RestartPartitionsCall.java:54]: I would expect the CLI command to complete when the response is returned, which happens when future completes. It seems that this has to be verified.

File: modules/cli/src/integrationTest/java/org/apache/ignite/internal/cli/commands/recovery/restart/ItRestartPartitionsTest.java (hunk 2 above)
Comment (rpuch) [reply modules/cli/src/integrationTest/java/org/apache/ignite/internal/cli/commands/recovery/restart/ItRestartPartitionsTest.java:96]: Yep, in this case it's enought to make sure that the needed parameters get passed  to each next layer.

File: modules/cli/src/main/java/org/apache/ignite/internal/cli/call/recovery/restart/RestartPartitionsCall.java (hunk 4 above)
Comment (Phillippko) [reply modules/cli/src/main/java/org/apache/ignite/internal/cli/call/recovery/restart/RestartPartitionsCall.java]: You're right, we wait for restart. I was confused because of snapshot commands (they complete when process is started, but only because future doesn't wait for it to finish). Fixed

File: modules/cli/src/integrationTest/java/org/apache/ignite/internal/cli/commands/recovery/restart/ItRestartPartitionsTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (Phillippko) [reply modules/cli/src/integrationTest/java/org/apache/ignite/internal/cli/commands/recovery/restart/ItRestartPartitionsTest.java]: Talked to Mikhail, we usually don't write such tests)
